src/osd/PGLog.cc

```diff
@@ -1007,8 +1013,12 @@ void PGLog::read_log(ObjectStore *store, coll_t pg_coll,
 	object_info_t oi(bv);
 	if (oi.version < i->version) {
 	  ldpp_dout(dpp, 15) << "read_log  missing " << *i
-			     << " (have " << oi.version << ")" << dendl;
+			     << " (have " << oi.version << ")"
+			     << " unmodified_omap " << i->unmodified_omap
+			     << " unmodified_extents " << i->unmodified_extents << dendl;
 	  missing.add(i->soid, i->version, oi.version);
+	  missing.missing[i->soid].unmodified_omap = i->unmodified_omap;
+	  missing.missing[i->soid].unmodified_extents.insert(i->unmodified_extents);
```

Comment:
This isn't correct.  Suppose the log is:

(10'5, 10'8]

with entries

10'6 modify foo unmodified_omap=true unmodified_extents=[0, 1024]
10'7 modify foo unmodified_omap=false unmodified_extents=[1024, 2048]
10'8 modify foo unmodified_omap=true unmodified_extents=[0, MAX]

If the on disk version if 10'6, this code will erroneously mark unmodified_omap=true and unmodified_extents=[0,MAX] causing recovery to *only recover the xattrs* even though it needs to also recover the omap and [0, 1024] of the data payload.  You need to walk the log from oi.version up to i->version and combine all of the log entries.  That logic should be encapsulated in PGLog and should be covered in the PGLog unit tests.  Those PGLog unit tests also need to be modified to test the logic for merging log entries with unmodified_omap and unmodified_extents into the missing entries.